### DISCUSSION TITLE
What's the best way to initialize structs like FWP_BYTE_ARRAY16?

### DISCUSSION THREAD

**Discussion by HavenDV** [microsoft/CsWin32#520] [Q&A] [answered]
When generating, I periodically encounter similar structures:
```cs
internal partial struct FWP_BYTE_ARRAY16
{
	/// <summary>Array of 16 bytes.</summary>
	internal __byte_16 byteArray16;

	internal partial struct __byte_16
	{
		internal byte _0,_1,_2,_3,_4,_5,_6,_7,_8,_9,_10,_11,_12,_13,_14,_15;

		/// <summary>Always <c>16</c>.</summary>
		internal readonly int Length => 16;
	}
}
```
What's the best way to initialize them, given that there are similar structures with 256 values?

Right now I am using something like this:
```cs
internal static FWP_BYTE_ARRAY16 ToArray16(this IPAddress address)
    {
        var bytes = address.GetAddressBytes();

        return new FWP_BYTE_ARRAY16
        {
            byteArray16 = new FWP_BYTE_ARRAY16.__byte_16
            {
                _0 = bytes[0],
                _1 = bytes[1],
                _2 = bytes[2],
                _3 = bytes[3],
                _4 = bytes[4],
                _5 = bytes[5],
                _6 = bytes[6],
                _7 = bytes[7],
                _8 = bytes[8],
                _9 = bytes[9],
                _10 = bytes[10],
                _11 = bytes[11],
                _12 = bytes[12],
                _13 = bytes[13],
                _14 = bytes[14],
                _15 = bytes[15],
            }
        };
    }
```

**Comment by HavenDV** [comment]
I suggest adding something like this to those structures
```cs
public unsafe void Set(string text)
{
    text = text ?? string.Empty;

    fixed (char* p0 = &_0)
    {
        for (var i = 0; i < text.Length && i < Length; i++)
        {
            p0[i] = text[i];
        }

        var nullTerminatedIndex = global::System.Math.Min(text.Length, Length - 1);
        p0[nullTerminatedIndex] = '\0';
    }
}
```

Now I am using
```cs
internal static class StringUtilities
{
    public static unsafe void SetTo(
        this string text,
        char* start,
        int maxLength)
    {
        for (var i = 0; i < text.Length && i < maxLength; i++)
        {
            start[i] = text[i];
        }

        start[Math.Min(text.Length, maxLength - 1)] = '\0';
    }

}

fixed (char* p0 = &data._0)
{
    title.SetTo(p0, data.Length);
}
```

**Comment by AArnott** [comment, marked as answer]
The `FWP_BYTE_ARRAY16` you refer to is generated with several helpers, depending on the language version and target framework of your project.

When targeting net472 and using C# 10, you get this:
```cs
internal unsafe readonly void CopyTo(Span<byte> target, int length = 16)
{
	if (length > 16) throw new ArgumentOutOfRangeException("length");
	fixed (byte* p0 = &_0) 
		for(int i = 0; i < length; i++) target[i]= p0[i];
}
```

We also offer these extension methods, that allow indexing in:

```cs
internal static unsafe ref readonly byte ReadOnlyItemRef(this in winmdroot.NetworkManagement.WindowsFilteringPlatform.FWP_BYTE_ARRAY16.__byte_16 @this, int index)
{
    fixed (byte* p0 = &@this._0)
        return ref p0[index];
}

internal static unsafe ref byte ItemRef(this ref winmdroot.NetworkManagement.WindowsFilteringPlatform.FWP_BYTE_ARRAY16.__byte_16 @this, int index)
{
    fixed (byte* p0 = &@this._0)
        return ref p0[index];
}
```

When targeting .NET Core 3.1 or later, you get that, and more:

```cs
/// <summary>
/// Gets a ref to an individual element of the inline array.
/// ⚠ Important ⚠: When this struct is on the stack, do not let the returned reference outlive the stack frame that defines it.
/// </summary>
internal ref byte this[int index] => ref AsSpan()[index];

/// <summary>
/// Gets this inline array as a span.
/// </summary>
/// <remarks>
/// ⚠ Important ⚠: When this struct is on the stack, do not let the returned span outlive the stack frame that defines it.
/// </remarks>
internal Span<byte> AsSpan() => MemoryMarshal.CreateSpan(ref _0, 16);
```

With the AsSpan and indexers, you can treat the fields with the same syntax as an array.

**Reply by HavenDV** [reply]
I may be missing something, but `0.1.647-beta` generates the following for net451/netstandard20:
```cs
// ------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
// ------------------------------------------------------------------------------

#pragma warning disable CS1591,CS1573,CS0465,CS0649,CS8019,CS1570,CS1584,CS1658,CS0436
namespace Windows.Win32
{
	using global::System;
	using global::System.Diagnostics;
	using global::System.Runtime.CompilerServices;
	using global::System.Runtime.InteropServices;
	using winmdroot = global::Windows.Win32;

	namespace NetworkManagement.WindowsFilteringPlatform
	{
		/// <summary>Stores an array of exactly 16 bytes.</summary>
		/// <remarks>
		/// <para>This data type  is useful for IPv6 addresses.</para>
		/// <para><see href="https://docs.microsoft.com/windows/win32/api//fwptypes/ns-fwptypes-fwp_byte_array16#">Read more on docs.microsoft.com</see>.</para>
		/// </remarks>
		internal partial struct FWP_BYTE_ARRAY16
		{
			/// <summary>Array of 16 bytes.</summary>
			internal __byte_16 byteArray16;

			internal partial struct __byte_16
			{
				internal byte _0,_1,_2,_3,_4,_5,_6,_7,_8,_9,_10,_11,_12,_13,_14,_15;

				/// <summary>Always <c>16</c>.</summary>
				internal readonly int Length => 16;
			}
		}
	}
}
```

**Reply by AArnott** [reply]
What is your `LangVersion` project property set to? Try setting it to 8, 9 or 10. I'm sure 9 is enough -- I'm not sure if 8 is.

**Reply by HavenDV** [reply]
```xml
<LangVersion>preview</LangVersion>
```
via Directory.Build.props

**Reply by HavenDV** [reply]
more precisely, there are two Directory.Build.props, one common for all projects (where the version is indicated), and one at the level below, but including `<Import Project="..\Directory.Build.props" />`

**Reply by AArnott** [reply]
I don't understand why you're not getting more. With net451 I'm still getting all the code I showed for net472.

Here is my project, which compiles fine. Does it work for you?
[CsWin32Sandbox.zip](https://github.com/microsoft/CsWin32/files/8637260/CsWin32Sandbox.zip)

**Reply by HavenDV** [reply]
[CsWin32Sandbox_Repro.zip](https://github.com/microsoft/CsWin32/files/8637315/CsWin32Sandbox_Repro.zip)
I'm not entirely sure what specifically leads to the problem. Either `net4.5.1` instead of `net451`, or a series of Directory.Build files.

**Reply by AArnott** [reply]
I've never seen `net4.5.1` used before. The supported syntax is just `net451`. Does switching to that fix it for you?

**Reply by jnm2** [reply]
`net4.5.1` is the officially recommended syntax now, I believe.

**Reply by HavenDV** [reply]
Please check the repro zip I sent you. I couldn't get additional methods to appear in FWP_BYTE_ARRAY16 just by changing the TargetFramework. It is possible that several conditions are required for the problem to occur. Maybe it's some kind of caching issue or VS IDE

**Reply by AArnott** [reply]
Your repro built just fine. But all your code was commented out. When I uncommented it, it failed to build because your project is a class library instead of an exe. When I add `<OutputType>exe</OutputType>` to the project file, it builds fine, and I can see the generated code I mentioned should exist.

![image](https://user-images.githubusercontent.com/3548/167157297-2f486cff-d863-4efa-86e3-2b3d550f4b7f.png)

**Reply by HavenDV** [reply]
I mean there are no additional methods in FWP_BYTE_ARRAY16

**Reply by AArnott** [reply]
The methods you get given your project are limited to the extension methods. 
When I add a PackageReference to System.Memory, I get `CopyTo`, `ToArray` and `Equals` added to the struct. This is odd because System.Memory should be coming down transitively because you reference CsWin32. So I did more digging, and found that applying this change to your project file fixes it:

```diff
 		<PackageReference Include="Microsoft.Windows.CsWin32" Version="0.1.647-beta">
 			<PrivateAssets>all</PrivateAssets>
-			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
 		</PackageReference>
```

The metadata you had on your PackageReference was suppressing the System.Memory dependency from being added to your compiler, thus making anything dealing with Span types off-limits, and limiting the code generation you could get.

**Reply by HavenDV** [reply]
This is the standard entry when installing a package via NuGet if `<DevelopmentDependency>true</DevelopmentDependency>` is specified.
Maybe it's worth noting this in the documentation?

**Reply by jnm2** [reply]
Also upvote https://github.com/NuGet/Home/issues/10746 since NuGet does not make it possible today for a development dependency to properly express adding package dependencies to the consuming project. IIRC.

**Reply by AArnott** [reply]
Good suggestions, both of you. Thank you. I'll update the readme to suggest removal of the `IncludeAssets` metadata.
